### PR TITLE
CDEC-415 Preprare for removal of `HasProtocolConstants` constraint

### PR DESCRIPTION
**Pull request by erikd**
## Description

The types `EpochOrSlot`, `LocalSlotIndex` and `SlotId` have a `HasProtocoConstants` constraint as well as `Enum` instances which depend on the `HasProtocoConstants` constraint. Since we want to remove `HasProtocoConstants` the first step is to provide functions that will act like the `Enum` methods, but take an extra parameter for the part they require from `ProtocolConstants`.

Also add property tests that show the equivalence of the two.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-415

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/Pos/Core/Slotting/LocalSlotIndex.hs [hunk 1]
```diff
@@ -95,12 +103,35 @@ addLocalSlotIndex x (UnsafeLocalSlotIndex i)
...
     s = fromIntegral x + fromIntegral i
 
--- | Unsafe constructor of 'LocalSlotIndex'.
-unsafeMkLocalSlotIndex :: HasProtocolConstants => Word16 -> LocalSlotIndex
-unsafeMkLocalSlotIndex = unsafeMkLocalSlotIndexExplicit protocolConstants
+unsafeMkLocalSlotIndexExplicit :: SlotCount -> Word16 -> LocalSlotIndex
+unsafeMkLocalSlotIndexExplicit es =
+    leftToPanic "unsafeMkLocalSlotIndex failed: " . mkLocalSlotIndexExplicit es
+
+
```
Comment (intricate) [inline core/src/Pos/Core/Slotting/LocalSlotIndex.hs]: Unnecessary newline here

File: core/test/Test/Pos/Core/EnumEmpTest.hs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import qualified Hedgehog.Range as Range
+
+import           Pos.Core (pcEpochSlots, withProtocolConstants)
+import           Pos.Core.Slotting (epochOrSlotFromEnum, epochOrSlotToEnum,
+                     localSlotIndexFromEnum, localSlotIndexToEnum,
+                     slotIdFromEnum, slotIdToEnum)
+
+import           Test.Pos.Core.Gen
+
+
```
Comment (intricate) [inline core/test/Test/Pos/Core/EnumEmpTest.hs]: Unnecessary newline here

File: core/test/Test/Pos/Core/EnumEmpTest.hs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
```
Comment (intricate) [inline core/test/Test/Pos/Core/EnumEmpTest.hs]: Unnecessary newline here

File: core/test/Test/Pos/Core/EnumEmpTest.hs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+prop_enum_eqivanent_epochOrSlotFromEnum :: Property
+prop_enum_eqivanent_epochOrSlotFromEnum =
+  H.withTests 5000 . H.property $ do
```
Comment (intricate) [inline core/test/Test/Pos/Core/EnumEmpTest.hs]: Shouldn't we indent by 4 spaces here or am I mistaken? (This applies to all the props)

Comment (erikd) [reply core/test/Test/Pos/Core/EnumEmpTest.hs]: Yes we should. Thanks.

File: core/test/Test/Pos/Core/EnumEmpTest.hs [hunk 5]
```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+module Test.Pos.Core.EnumEmpTest
+       ( tests
+       ) where
+
+-- The types `EpochOrSlot`, `LocalSlotIndex` and `SlotId` have a
+-- `HasProtocoConstants` constraint as well as `Enum` instances which depend
+-- on the `HasProtocoConstants` constraint. Since we want to remove
+-- `HasProtocoConstants` the first step is to provide functions that will act
```
Comment (mhuesch) [inline core/test/Test/Pos/Core/EnumEmpTest.hs]: typo: `HasProtocoConstants` should be `HasProtocolConstants`

File: core/test/Test/Pos/Core/EnumEmpTest.hs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        new <- catchErrorCall $ slotIdToEnum (pcEpochSlots pc) i
+        old === new
+
+-- | Evaluate and force a pure value and if the evaluation of the pure value
+-- causes 'error' to be called, catch 'ErrorCall' and return the error message
+-- with the "functionName:" part removed.
+--
+-- This is the stupid level of crap you have to go through when people
+-- think its ok to sprinkle 'error' and 'MonadError' throughout otherwise
+-- pure code.
```
Comment (parsonsmatt) [inline core/test/Test/Pos/Core/EnumEmpTest.hs]: :fire: :fire_engine:

**Pull request merged**